### DISCUSSION TITLE
How to keep atom store size under control

### DISCUSSION THREAD

**Discussion by mrodyhin** [pmndrs/jotai#2312] [Q&A]
Hi, i'm pretty new to Jotai, and i would really appreciate if someone can help me to better understand how to use it properly, and correct me if my understanding is incorrect.

All my concerns starts from that line in the documentation:
>When an atom is no longer used, meaning all the components using it are unmounted, **and the atom config no longer exists**, the value is removed from the Provider.

If i understand it correct, that means, every time i create a new atom in the module scope, for example like that:
`export const myAtom = atom();`
and then use it, it will keep it last state in the store "forever" since it's config can not be removed.
I found a few ways that can partially resolve that:
1) using `atomWithReset`, and manually resetting  atom data when needed
2) using `onUnmount` and automatically set empty data when atom looses all it subscribers
Neither of that actually remove the atom from state, and both adds additional overhead, so let me know if there is better solution for that.

But what really warns me, is using of the Jotai Query. Here is the basic example rom the docs:
```ts
import { atom, useAtom } from 'jotai'
import { atomWithQuery } from 'jotai-tanstack-query'

const idAtom = atom(1)
const userAtom = atomWithQuery((get) => ({
  queryKey: ['users', get(idAtom)],
  queryFn: async ({ queryKey: [, id] }) => {
    const res = await fetch(`https://jsonplaceholder.typicode.com/users/${id}`)
    return res.json()
  },
}))

const UserData = () => {
  const [{ data, isPending, isError }] = useAtom(userAtom)

  if (isPending) return <div>Loading...</div>
  if (isError) return <div>Error</div>

  return <div>{JSON.stringify(data)}</div>
}
```
If `userAtom` in that example behaves the same way as the regular atom, it's data won't be removed from store after `UserData` component would be unmouted. And it will be even worse, if derived atoms will be involved.
I'm really concerned about that, because application can contain a lot of such atoms, and that atoms can hold significant amount of data.
One way to avoid that can be to create it inside of the component, or wrap it in hook, for example like that:
```ts
const useUserAtom = () => {
  const userAtom = useMemo(() => atomWithQuery((get) => ({
    queryKey: ['users', get(idAtom)],
    queryFn: async ({ queryKey: [, id] }) => {
      const res = await fetch(`https://jsonplaceholder.typicode.com/users/${id}`)
      return res.json()
    },
  })), []);
  return userAtom;
}
```
but i don't like that, because:

- it's create a lot of additional code
- it significantly complicates working with derived atoms that needs to use that atom
- using that in multiple places, will create multiple atoms which are basically the same.

So i'm looking for a better solution that will help to keep store size under control.

**Comment by dai-shi** [comment]
> If i understand it correct, that means, every time i create a new atom in the module scope, for example like that:
> `export const myAtom = atom();`
> and then use it, it will keep it last state in the store "forever" since it's config can not be removed.

Yeah, to remove atom from the module scope, you need to explicitly do it. This is a _naive_ way:
```js
export let myAtom = atom();
export const destroyMyAtom = () => { myAtom = undefined; };
```

It's the same as you define a big object in a module scope.

**Comment by mrodyhin** [comment]
Yes but when everything is reactive, doing it in a such manual way looks strange. I'm kind of shocked that people uses atomWithQuery like what without any concerns.
Don't you have any ideas how to implement auto remove, when nothing is using the atom?

I imagine it could be something like that:
```js
const atomWithAutoDestroy = (createAtom) => {
  let atomLink;
  let _atom = ((get) => get(atomLink));
  _atom.onMount = () => {
    atomLink = createAtom();
    return () => { atomLink = undefined }
  }
  return _atom;
}
// usage
const myAtom = atomWithAutoDestroy(() => atom('possibly lots of data'))
```
But i'm not sure how to correctly make it work with writable atoms, and is everything would be ok with derived atoms of myAtom.

**Reply by mrodyhin** [reply]
Just realised that it probably won't work since _atom is subscribed to atomLink, and atomLink atom won't be destroyed even if i will rewrite it.

**Reply by dai-shi** [reply]
To clarify, Jotai core doesn't hold anything after unused. It's outside of Jotai core, if something is held. So, I think it's the responsibility of 'jotai-tanstack-query' in your case.

**Reply by mrodyhin** [reply]
But if i create a regular atom in module scope, than in component set some value into that, i'll still see that value even after remounting that component.

**Reply by mrodyhin** [reply]
And i want to achieve the opposite. I want atom to be able to clear it's data in the store, when it looses all subscribers.

**Reply by dai-shi** [reply]
The lifetime of atom in module scope has nothing to do with React lifecycle.
Are you aware that using `<Provider>` you can clear all data on unmount?

**Reply by mrodyhin** [reply]
>The lifetime of atom in module scope has nothing to do with React lifecycle.

I agree, and i don want it to depend on react. I want it to depend on atom itself, and it's number of subscribers. It looks like https://github.com/pmndrs/jotai/issues/84#issuecomment-703925391 Jotai even had something like that back in a days.

>Are you aware that using \<Provider\> you can clear all data on unmount?

You probably mean that i can use multiple stores, and destroy the whole store when it's not needed. But if i understand it correctly, stores are completely independent, so if atom would be used in 2 different provider scopes (even nested), it would be duplicated, and it also won't be in sync. That means, it's not good to create small scopes, because of that duplications, and it's not good to create large scopes, because it's only possible to destroy it completely.

**Reply by dai-shi** [reply]
Back in days, we used to deleted atom values if subscribers become empty. I think it's before v1, and that was troublesome for some use cases, and users had no control. So, we decided to make atom values lifetime as the same as atom lifetime, which is a big design principle of Jotai.

Maybe you want to try something like this:
```js
const initialValue1 = ...;
const atom1 = atom(initialValue1);
const initialValu2 = ...;
const atom2 = atom(initialValue2);

const catchAllAtom = atom(
  null,
  (get, set, action) => {
    if (action === 'reset') {
      set(atom1, initialValue1);
      set(atom2, initialValue2);
    }
  },
);
catchAllAtom.onMount = (set) => {
  const cleanup = () => set('reset');
  return cleanup;
};
```

**Reply by mrodyhin** [reply]
> Back in days, we used to deleted atom values if subscribers become empty.

I definitely agree, that having no control over that was not good, but if we think, there is no much more control now.
As for me, the idea that something is not needed when it's not used anymore is definitely useful for lots of cases, and it seems like it's not really easy to implement it with the current approach.

**Reply by dai-shi** [reply]
How about creating a wrapper on your end?

```ts
import { atom as atomOrig } from 'jotai';

export const atom: typeof atomOrig = (...args) => {
  const a = atomOrig(...args);
  if (a.write) {
    a.onMount = (set) => {
      return () => set(null);
    };
  }
  return a;
};
```

Hmm, I don't like this ☝️  approach very much.

```js
import { atom as atomOrig } from 'jotai';

const atomSet = new Set();

export const atom = (...args) => {
  const a = atomOrig(...args);
  atomSet.add(a);
  return a;
};

export const clearAtoms = (store) => {
  for (const a of atomSet) {
    store.set(a, null);
  }
  atomSet.clear();
};
```

From the library perspective, atom state is created with `store.get()` not `store.sub()`, so we can't really clear it on unsubscribe.
